spec/AuthenticationAdapters.spec.js

```diff
@@ -897,6 +898,183 @@ describe('keycloak auth adapter', () => {
+    try {
+      await auth0.validateAuthData({ id_token: 'token' }, { clientId: 'secret' });
+      fail();
+    } catch (e) {
+      expect(e.message).toBe('tenant id is invalid.');
+    }
+  });
+
+  it('Should throw error with missing client_id', async () => {
+    try {
+      await auth0.validateAuthData({ id_token: 'token' }, { tenantId: 'example.eu.auth0.com' });
+      fail();
+    } catch (e) {
+      expect(e.message).toBe('client id is invalid.');
+    }
+  });
+
+  it('should not decode invalid id_token', async () => {
+    try {
+      await auth0.validateAuthData(
+        { id: 'the_user_id', id_token: 'the_token' },
+        { tenantId: 'example.eu.auth0.com', clientId: 'secret' }
+      );
+      fail();
+    } catch (e) {
+      expect(e.message).toBe('provided token does not decode as JWT');
+    }
+  });
+
+  it('should throw error if public key used to encode token is not available', async () => {
+    const fakeDecodedToken = { header: { kid: '789', alg: 'RS256' } };
+    try {
+      spyOn(jwt, 'decode').and.callFake(() => fakeDecodedToken);
+
+      await auth0.validateAuthData(
+        { id: 'the_user_id', id_token: 'the_token' },
+        { tenantId: 'example.eu.auth0.com', clientId: 'secret' }
+      );
+      fail();
+    } catch (e) {
+      expect(e.message).toBe(
+        `Unable to find matching key for Key ID: ${fakeDecodedToken.header.kid} for auth0 tenantId example.eu.auth0.com.`
+      );
+    }
+  });
+
+  it('should use algorithm from key header to verify id_token', async () => {
+    const fakeClaim = {
+      iss: 'https://example.eu.auth0.com/',
+      aud: 'secret',
+      exp: Date.now(),
+      sub: 'the_user_id',
+    };
+    const fakeDecodedToken = { header: { kid: '123', alg: 'RS256' } };
+    spyOn(jwt, 'decode').and.callFake(() => fakeDecodedToken);
+    spyOn(jwt, 'verify').and.callFake(() => fakeClaim);
+    const fakeGetSigningKeyAsyncFunction = () => {
+      return { kid: '123', rsaPublicKey: 'the_rsa_public_key' };
+    };
+    spyOn(util, 'promisify').and.callFake(() => fakeGetSigningKeyAsyncFunction);
+
+    const result = await auth0.validateAuthData(
+      { id: 'the_user_id', id_token: 'the_token' },
+      { tenantId: 'example.eu.auth0.com', clientId: 'secret' }
+    );
+    expect(result).toEqual(fakeClaim);
+    expect(jwt.verify.calls.first().args[2].algorithms).toEqual(fakeDecodedToken.header.alg);
+  });
+
+  it('should not verify invalid id_token', async () => {
+    const fakeDecodedToken = { header: { kid: '123', alg: 'RS256' } };
+    spyOn(jwt, 'decode').and.callFake(() => fakeDecodedToken);
+    const fakeGetSigningKeyAsyncFunction = () => {
+      return { kid: '123', rsaPublicKey: 'the_rsa_public_key' };
+    };
+    spyOn(util, 'promisify').and.callFake(() => fakeGetSigningKeyAsyncFunction);
+
+    try {
+      await auth0.validateAuthData(
+        { id: 'the_user_id', id_token: 'the_token' },
+        { tenantId: 'example.eu.auth0.com', clientId: 'secret' }
```

Comment:
Is this a URI that is made available for development use per policy?

Reply:
Not, as far as I know, I just choose a value for the tests. Since the tenant is not really used in the tests, it does not really matter. Another option would be to create an auth0 account and register a tenant, but then someone from the Parse team would have to manage this account.

Reply:
Is there any benefit regarding test efficacy if we create an auth0 account? If not, I suggest we comply with RFC 2606 and replace the URLs in the tests with `example.com`.